Only load the javascript needed for a specific view (esp. hips.js, google maps, ckeditor)

weedySeaDragon opened the pull request:
PT Story: 
https://www.pivotaltracker.com/story/show/153991394

Changes proposed in this pull request:
1. added hooks into application layout to load meta-info, stylesheets and javascript specific to a view (using yield statements)
   - went ahead and added in the hooks for meta-info and stylesheets because those are chores TBD; was easy to do it now to establish the pattern
2. defined a step to use in features to check for a javascript file in the header (just have to match a "contains(...)" so you only have to know part of the javascript file or path
3. created a feature describing the goal and testing for it

Ready for review:
@AgileVentures/shf-project-team

weedySeaDragon committed:
+ step: look for <javascript> in the head section where the src: attribute contains something

weedySeaDragon committed:
hooks in the layout to put in page-specific meta info, styles, javascript

weedySeaDragon committed:
feature: only load javascript needed for the particular page

weedySeaDragon committed:
load javascript needed for the specific pages (views) via content_for :javascript_for_view

patmbolger commented on features/page-loads-only-js-needed.feature:
```diff
@@ -0,0 +1,146 @@
+    Given I am logged in as "applicant@bowwow.se"
+    And I am on the "user details" page for "applicant@bowwow.se"
+    Then the page source should have "hips.js" javascript in the header
+    And the page source should not have "maps.googleapis.com" javascript in the header
```
want to check for (not) creditor as well

patmbolger commented on app/views/companies/show.html.haml:
```diff
@@ -1,3 +1,7 @@
+- content_for :javascript_for_view do
+  = javascript_include_tag "https://maps.googleapis.com/maps/api/js?key=#{Geocoder.config[:api_key]}&language=#{I18n.locale}&region=SE"
+
+
```
We don't need this is this file.

We _do_ need it in `payments/create.html.haml` (payment for is not showing up right now as a result).

weedySeaDragon replied:
@patmbolger  Um, we _do_ need the google maps javascript.  (Perhaps your mind was just so totally on the payments stuff, you misread this?)
But this did prompt me to put scenarios into the feature file.

And I'll add the Hips javascript stuff  into `payments/create.html.haml`

But when you say
> (payment for is not showing up right now as a result).

Do you mean that (a) it's not visible to the user somewhere?  Or (b) that the HIPS_JS_CDN must be included in `payments/create.html.haml` so that the payment can be processed?
I'm assuming _(b)_ but I'm not totally sure that's what you mean.

weedySeaDragon replied:
Also --

What test (spec or feature) needs to be written so that we are clear on what should happen _and_ are testing for it?

If you give me the general idea, I'll write it.

patmbolger replied:
Oops ... so sorry to confuse the issue.  I meant for this comment to apply to `users.show.html.haml`.

patmbolger replied:
And, my comment should have been `payment form is not showing up right now`.

I need to slow down when I'm entering these comments and confirm they're correct !

patmbolger replied:
Good question re testing .... this approach (cucumber) seems like the hard way to conduct the tests _and_ to maintain the tests as we change where JS assets are loaded (and, perhaps in the future, CSS).

Another approach would be to use rspec view tests.  For instance:

`Confirm certains assets NOT loaded on every page`
  .... here, test that these assets are not loaded via `layouts/application.html.haml`.

`Confirm Ckeditor assets loaded on appropriate pages`

etc ....

patmbolger reviewed (changes requested):
Looks good - please see comments.

weedySeaDragon committed:
simplified features; add scenarios for company view/show page; landing should not have ckeditor.js

weedySeaDragon committed:
add HIPS_JS_CDN javascript to payment/create view

weedySeaDragon committed:
Merge branch 'develop' into sprint33-javascript-per-page-#153991394

patmbolger commented on app/views/users/show.html.haml:
```diff
@@ -1,3 +1,6 @@
+- content_for :javascript_for_view do
+  = javascript_include_tag HIPS_JS_CDN
```
_This_ is where I meant my previous comment to go ....

patmbolger reviewed (changes requested):
Updated comments.

weedySeaDragon committed:
Merge branch 'develop' into sprint33-javascript-per-page-#153991394

weedySeaDragon committed:
re-org payment locale entries to fit with conventions; update references to them in helpers

weedySeaDragon committed:
Merge branch 'develop' into sprint33-javascript-per-page-#153991394

weedySeaDragon committed:
be explicit about what is expected in the view; change date because real date change caused different data displayed

Time passing in the real world meant that the words displayed in the "created date" had changed.  I changed the expectations to allow for that.

weedySeaDragon committed:
add steps: hide (or show) search form, scroll to top (same as PR 459; also needed here)

weedySeaDragon committed:
add variation to "I should see.." step: case insensitive.  Needed to check page titles

weedySeaDragon committed:
+Scenario: different page titles for roles; sort displayed data; feature title, hints,meaningful data/var names;

weedySeaDragon committed:
simplify conditional to show title

thesuss reviewed (approved):
If Patricks comment is sorted

weedySeaDragon committed:
Merge branch 'develop' into sprint33-javascript-per-page-#153991394

# Conflicts:
#	features/manage_users.feature

thesuss commented:
Here's a loot of conflicting files now. How about merge the redesign and then look into this one again?

patmbolger commented:
Closing for now as this is not urgent but may be a nice-to-have in the future.
